### PR TITLE
neonsecret attention optimization

### PR DESCRIPTION
**Pull request by neonsecret**

### CODE REVIEW

**Comment by harubaru** [comment]
What does this do?

**Comment by neonsecret** [comment]
practically better attention module, better vram usage, see https://github.com/neonsecret/stable-diffusion

**Comment by harubaru** [comment]
@neonsecret How much VRAM does this save for inference and training usecases, and how fast is inference? I read your README but it doesn't go into specifics on what hardware was used, especially for a usecase such as training.

**Comment by neonsecret** [comment]
> @neonsecret How much VRAM does this save for inference and training usecases, and how fast is inference? I read your README but it doesn't go into specifics on what hardware was used, especially for a usecase such as training.

it does

**Comment by chavinlo** [comment]
@neonsecret How much VRAM did you managed to save? I tried running training but rather than decreasing the amount of VRAM in use, it increases it. From 30.5GB to 35.5GB.

**Comment by neonsecret** [comment]
> @neonsecret How much VRAM did you managed to save? I tried running training but rather than decreasing the amount of VRAM in use, it increases it. From 30.5GB to 35.5GB.
1. what version of my fork did you use?
2. what params did you set?
3. what commands did you use?
4. how did you install the repo?

**Comment by neonsecret** [comment]
> > > @neonsecret How much VRAM did you managed to save? I tried running training but rather than decreasing the amount of VRAM in use, it increases it. From 30.5GB to 35.5GB.
> > 
> > 
> > 
> > 1. what version of my fork did you use?
> > 2. what params did you set?
> > 3. what commands did you use?
> > 4. how did you install the repo?
> 
> 1. I just patched the files from this repo with the ones of this PR
> 2. default 4gpu yaml
> 3. python3 main.py -t -n "waifu-diffusion" --resume_from_checkpoint model.ckpt --base ./configs/stable-diffusion/v1-finetune-4gpu.yaml --no-test --seed 25 --scale_lr False --gpus 0, --data_root "./danbooru-aesthetic"
> 4. just cloned, install dependencies, patch checkpoint_connector.py, and run with command above
it definitely shouldn't increase memory usage, so perhaps you did something wrong with measurements or there's been a memory leak at your system

**Comment by neonsecret** [comment]
> > > @neonsecret How much VRAM did you managed to save? I tried running training but rather than decreasing the amount of VRAM in use, it increases it. From 30.5GB to 35.5GB.
> > 
> > 
> > 
> > 1. what version of my fork did you use?
> > 2. what params did you set?
> > 3. what commands did you use?
> > 4. how did you install the repo?
> 
> 1. I just patched the files from this repo with the ones of this PR
> 2. default 4gpu yaml
> 3. python3 main.py -t -n "waifu-diffusion" --resume_from_checkpoint model.ckpt --base ./configs/stable-diffusion/v1-finetune-4gpu.yaml --no-test --seed 25 --scale_lr False --gpus 0, --data_root "./danbooru-aesthetic"
> 4. just cloned, install dependencies, patch checkpoint_connector.py, and run with command above

maybe there was a memory leak at your system, the optimizations definitely shouldn't decrease memory usage

**Comment by harubaru** [comment]
The only modifications done to the original source code are to fix bugs with the trainer and to add a custom dataloader. There are no memory leaks with the vanilla attention during training.
